honeycombio/data_source_query.go

```diff
@@ -284,3 +284,71 @@ func validateQueryJSON(validators ...querySpecValidateDiagFunc) schema.SchemaVal
 		return diagnostics
 	}
 }
+
+func extractFilters(d *schema.ResourceData) ([]honeycombio.FilterSpec, error) {
+	filterSet := d.Get("filter").(*schema.Set)
+	hashFn := filterSet.F
+	filters := make([]honeycombio.FilterSpec, filterSet.Len())
+	for i, schemaFilter := range filterSet.List() {
+		honeyFilter, err := extractFilter(d, hashFn(schemaFilter))
+		if err != nil {
+			return nil, err
+		}
+		filters[i] = honeyFilter
+	}
+	return filters, nil
+}
+
+const multipleValuesError = "must choose one of 'value', 'value_string', 'value_integer', 'value_float', 'value_boolean'"
+
+func extractFilter(d *schema.ResourceData, id int) (honeycombio.FilterSpec, error) {
```

Comment:
add a comment on the method/maybe explain the mutually exclusive value stuff

Reply:
To investigate: the [`Schema`](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/v2@v2.0.1/helper/schema?tab=doc#Schema) type of the Terraform SDK has a number of properties that might be able to do this validation for us:

```
	// ConflictsWith is a set of schema keys that conflict with this schema.
	// This will only check that they're set in the _config_. This will not
	// raise an error for a malfunctioning resource that sets a conflicting
	// key.
	//
	// ExactlyOneOf is a set of schema keys that, when set, only one of the
	// keys in that list can be specified. It will error if none are
	// specified as well.
	//
	// AtLeastOneOf is a set of schema keys that, when set, at least one of
	// the keys in that list must be specified.
	//
	// RequiredWith is a set of schema keys that must be set simultaneously.
	ConflictsWith []string
	ExactlyOneOf  []string
	AtLeastOneOf  []string
	RequiredWith  []string
```

I haven't used these before, so it's not clear to me how they work and if they will fit our use case.

Reply:
Good catch, but it looks like it isn't supported for sets: https://github.com/hashicorp/terraform-plugin-sdk/issues/71

Tried the following:
 
`ConflictsWith: []string{"value_string", "value_integer", "value_float", "value_boolean"},`
->
```
      	* data source honeycombio_query: ConflictsWith: value references unknown
        attribute (value_string) at part (value_string)
```

---

`ConflictsWith: []string{"filter.0.value_string", "filter.0.value_integer", "filter.0.value_float", "filter.0.value_boolean"},`
->
```
        	* data source honeycombio_query: ConflictsWith: value_float configuration
        block reference (filter.0.value) can only be used with TypeList and MaxItems:
        1 configuration blocks
```

---

`ConflictsWith: []string{".value_string", ".value_integer", ".value_float", ".value_boolean"},`
->
```
        	* data source honeycombio_query: ConflictsWith: value_string references
        unknown attribute (.value) at part ()
```